File: etc/lua/tag/WmDropmenu.lua
local Class = require("util.Class")
local Tag = require("engine.Tag")
local WmDropmenu = Class(Tag)

--htmlbuff
function WmDropmenu:createNew(obj,htmlbuff)
  self.htmlbuff = htmlbuff
  self.actions = {}
end

--attr所有属性
function WmDropmenu:doStartTag(attr)
  --属性初始化
  attr.closeAction = attr.closeAction
  attr.openAction = attr.oppenAction
  self.attr = attr
  self.htmlbuff:append('<div id="')
  self.htmlbuff:append(attr.id)
  self.htmlbuff:append('">','\n')
  self.htmlbuff:append([[
  	<div class="fn">
		<span class="e_button"><span>]])
  self.htmlbuff:append(attr.label)
  self.htmlbuff:append([[</span><span class="e_ico-unfold"></span></span>
	</div>
	<div class="c_dropmenu">
      	<ul>
    ]])
end

function WmDropmenu:doEndTag()
  self.htmlbuff:append([[
     	</ul>
      </div>
    </div>
  ]])
    
  self.htmlbuff:append([[
<script type="text/javascript">
  require(["wmDropmenu"],function(WmDropmenu) {
      var wmDropmenu=new WmDropmenu("]])
  self.htmlbuff:append(self.attr.id)
  self.htmlbuff:append('");','\n')
  self.htmlbuff:append([[
  	wmDropmenu.create();
  ]])
  
  if self.attr.openAction then
  	self.htmlbuff:append([[
      wmDropmenu.setOpenAction(function(){
    ]])
	self.htmlbuff:append(self.attr.openAction,'\n')
	self.htmlbuff:append('});','\n')
  end
  if self.attr.closeAction then
  	self.htmlbuff:append([[
	  wmDropmenu.setCloseAction(function(){
    ]])
 	self.htmlbuff:append(self.attr.closeAction,'\n')
	self.htmlbuff:append('});','\n')
  end
 
 self.htmlbuff:append('		var items = wmDropmenu.getDropmenuItems();','\n')
 for k,v in pairs(self.actions) do
    if v~=-1 then
      self.htmlbuff:append('items[')
      self.htmlbuff:append(k-1)
      self.htmlbuff:append('].setAction(function(){')
      self.htmlbuff:append(v)
      self.htmlbuff:append('});','\n')
    end
  end
  self.htmlbuff:append([[
	});
	</script>
  ]])
end

function WmDropmenu:addActions(action)
  table.insert(self.actions, action)
end

return WmDropmenu

--[[
require(["wmDropmenu","util"],function(WmDropmenu) {
  var wmDropmenu=new WmDropmenu(dropmenu01);
  wmDropmenu.create();
  wmDropmenu.setLabel("下拉菜单01");
  wmDropmenu.setCloseAction(function(){
    console.log("closeAction");
  });
  wmDropmenu.setOpenAction(function(){
    console.log("openAction");
    console.log(wmDropmenu.getLabel());
  });
  var items=wmDropmenu.getItems();
  for(var i=0;i<items.length;i++){
    console.log("index:"+items[i].getIndex());
    console.log("parent:"+items[i].getParent());
    items[i].setAction(function(obj){
      alert(obj.html());
    });
  }
  console.log("length:"+items.length);
  items[2].remove();
  console.log("length:"+items.length);
  for(var i=0;i<items.length;i++){
    //alert(items[i].html());
  }
  //wmDropmenu.removeAll();
  var item1=wmDropmenu.push("111");
  item1.setAction(function(obj){
    alert("我是新来的1");
  });
  var item2=wmDropmenu.push("22222");
  item2.setAction(function(obj){
    alert("我是新来的2");
  });
  item1.remove();
  //wmDropmenu.getItems()[0].remove();
  $("#testbtn").setAction(function(){
    if(wmDropmenu.Invisible()){
      wmDropmenu.show();
    }else{
      wmDropmenu.hidden();
    }
  });
});

]]

File: web/biz/js/tag/WmDropmenu.js
require(["domReady!","mobile","util"],function(doc,Mobile){
	new iScroll("content",{tap:true,scrollbars: true,useTransform:false});//增加页面滑动效果
	$("#testSwitch").tap(function() {
		var dropmenu = WmWebUI.select("myDropmenu");
		var that = this;
		if (dropmenu.invisible()) {
			dropmenu.show();
			$(that).html("关闭菜单");
		} else {
			dropmenu.hidden();
			$(that).html("打开菜单");
		}
	});
	
	$("#testSetLabel").tap(function(){
		var dropmenu = WmWebUI.select("myDropmenu");
		var that = this;
		if(dropmenu.getLabel()){
			dropmenu.setLabel(null);
			$(that).html("设置文本");
		}else{
			dropmenu.setLabel("自定义下拉菜单");
			$(that).html("清除文本");
		}
	});
	
	$("#testSetOpenAction").tap(function(){
		var dropmenu = WmWebUI.select("myDropmenu");
		var that = this;
		if(dropmenu.openAction){
			dropmenu.setOpenAction(null);
			$(that).html("设置打开事件");
		}else{
			dropmenu.setOpenAction(function(){
				alert("打开事件");
			});
			$(that).html("清除打开事件");
		}
	});
	
	$("#testSetCloseAction").tap(function(){
		var dropmenu = WmWebUI.select("myDropmenu");
		var that = this;
		if(dropmenu.closeAction){
			dropmenu.setCloseAction(null);
			$(that).html("设置关闭事件");
		}else{
			dropmenu.setCloseAction(function(){
				alert("关闭事件");
			});
			$(that).html("清除关闭事件");
		}
	});
	
	$("#testSetItemAction").tap(function() {
		var dropmenu = WmWebUI.select("myDropmenu");
		var items = dropmenu.getDropmenuItems();
		var that = this;
		if(items[0]&&items[0].action){
			for(var i=0;i<items.length;i++){
				items[i].setAction(null);
			}
			$(that).html("设置菜单项事件");
		}else{
			for(var i=0;i<items.length;i++){
				items[i].setAction(new Function("alert('第"+(i+1)+"个菜单项点击事件');"));
			}
			$(that).html("清空菜单项事件");
		}
	});
	/*清除所有菜单项*/
	$("#testRemoveAll").tap(function() {
		var dropmenu = WmWebUI.select("myDropmenu");
		dropmenu.removeAll();
		alert("清除成功");
	});
	/*添加菜单项*/
	$("#testPush").tap(function() {
		var dropmenu = WmWebUI.select("myDropmenu");
		dropmenu.push("下拉菜单项"+(dropmenu.getDropmenuItems().length+1));
		alert("添加成功")
	});
	/*删除菜单项*/
	$("#testRemove").tap(function() {
		var dropmenu = WmWebUI.select("myDropmenu");
		var item = dropmenu.getDropmenuItem(dropmenu.getDropmenuItems().length-1);
		item.remove();
		alert("删除成功");
	});
	/*更新菜单项*/
	$("#testHtml").tap(function(){
		var dropmenu=WmWebUI.select("myDropmenu");
		var item = dropmenu.getDropmenuItem(dropmenu.getDropmenuItems().length-1);//获取最后一个菜单项
		item.html("新的下拉菜单项");
		alert("更新成功");
	});
});